File: QueenMama/Services/Auth/AuthenticationManager.swift
import Foundation
import Combine

/// Main authentication coordinator for the macOS app
/// Manages login flows, token refresh, and authentication state
@MainActor
final class AuthenticationManager: ObservableObject {
    static let shared = AuthenticationManager()

    // MARK: - Published State

    @Published private(set) var authState: AuthState = .unknown
    @Published private(set) var isAuthenticated: Bool = false
    @Published private(set) var currentUser: AuthUser?

    // MARK: - Dependencies

    private let api = AuthAPIClient.shared
    private let tokenStore = AuthTokenStore.shared
    private let deviceInfo = DeviceInfo.current()

    // MARK: - Device Code Polling

    private var pollingTask: Task<Void, Never>?
    private let pollingInterval: TimeInterval = 5

    private init() {}

    // MARK: - Initialization

    /// Check for existing authentication on app launch
    func checkExistingAuth() async {
        authState = .unknown

        // Check for stored credentials
        guard tokenStore.hasStoredCredentials,
              let storedUser = tokenStore.storedUser else {
            authState = .unauthenticated
            return
        }

        // Try to validate/refresh tokens
        do {
            if !tokenStore.isAccessTokenValid {
                // Need to refresh
                guard let refreshToken = tokenStore.refreshToken else {
                    throw AuthError.tokenExpired
                }

                let response = try await api.refreshTokens(refreshToken)
                tokenStore.accessToken = response.accessToken
                tokenStore.accessTokenExpiry = Date().addingTimeInterval(TimeInterval(response.expiresIn))
                tokenStore.refreshToken = response.refreshToken
            }

            // Successfully authenticated
            currentUser = storedUser
            isAuthenticated = true
            authState = .authenticated(user: storedUser)

        } catch {
            print("[Auth] Existing auth validation failed: \(error)")
            // Clear invalid credentials
            tokenStore.clearAll()
            authState = .unauthenticated
        }
    }

    // MARK: - Credentials Login

    /// Login with email and password
    func loginWithCredentials(email: String, password: String) async throws {
        authState = .authenticating

        do {
            let response = try await api.login(
                email: email,
                password: password,
                deviceInfo: deviceInfo
            )

            // Store tokens
            let tokens = AuthTokens(
                accessToken: response.accessToken,
                refreshToken: response.refreshToken,
                expiresIn: response.expiresIn
            )
            tokenStore.storeTokens(tokens, user: response.user)

            // Update state
            currentUser = response.user
            isAuthenticated = true
            authState = .authenticated(user: response.user)

        } catch {
            authState = .error(message: error.localizedDescription)
            throw error
        }
    }

    // MARK: - Device Code Flow

    /// Start the device code flow for OAuth users
    func startDeviceCodeFlow() async throws -> DeviceCodeResponse {
        authState = .authenticating

        do {
            let response = try await api.requestDeviceCode(deviceInfo: deviceInfo)

            let expiresAt = Date().addingTimeInterval(TimeInterval(response.expiresIn))
            authState = .deviceCodePending(
                userCode: response.userCode,
                deviceCode: response.deviceCode,
                expiresAt: expiresAt
            )

            // Start polling
            startPolling(deviceCode: response.deviceCode, expiresAt: expiresAt)

            return response

        } catch {
            authState = .error(message: error.localizedDescription)
            throw error
        }
    }

    /// Cancel the device code flow
    func cancelDeviceCodeFlow() {
        pollingTask?.cancel()
        pollingTask = nil
        authState = .unauthenticated
    }

    private func startPolling(deviceCode: String, expiresAt: Date) {
        pollingTask?.cancel()

        pollingTask = Task {
            while !Task.isCancelled && Date() < expiresAt {
                do {
                    try await Task.sleep(nanoseconds: UInt64(pollingInterval * 1_000_000_000))

                    if Task.isCancelled { break }

                    let response = try await api.pollDeviceCode(deviceCode)

                    // Check if authorized
                    if let accessToken = response.accessToken,
                       let refreshToken = response.refreshToken,
                       let expiresIn = response.expiresIn,
                       let user = response.user {

                        // Store tokens
                        let tokens = AuthTokens(
                            accessToken: accessToken,
                            refreshToken: refreshToken,
                            expiresIn: expiresIn
                        )

                        await MainActor.run {
                            tokenStore.storeTokens(tokens, user: user)
                            currentUser = user
                            isAuthenticated = true
                            authState = .authenticated(user: user)
                        }

                        return
                    }

                    // Still pending - continue polling
                    if response.error == "authorization_pending" {
                        continue
                    }

                    // Some other error
                    if let error = response.error, error != "authorization_pending" {
                        await MainActor.run {
                            authState = .error(message: response.message ?? error)
                        }
                        return
                    }

                } catch {
                    // Ignore network errors during polling, just continue
                    print("[Auth] Polling error: \(error)")
                }
            }

            // Expired without authorization
            if !Task.isCancelled {
                await MainActor.run {
                    authState = .error(message: "Device code expired. Please try again.")
                }
            }
        }
    }

    // MARK: - Logout

    /// Logout from current device
    func logout() async {
        pollingTask?.cancel()

        do {
            try await api.logout(
                refreshToken: tokenStore.refreshToken,
                allDevices: false
            )
        } catch {
            print("[Auth] Logout API call failed: \(error)")
            // Continue with local logout anyway
        }

        tokenStore.clearAll()
        currentUser = nil
        isAuthenticated = false
        authState = .unauthenticated
    }

    /// Logout from all devices
    func logoutAllDevices() async {
        pollingTask?.cancel()

        do {
            try await api.logout(
                refreshToken: tokenStore.refreshToken,
                allDevices: true
            )
        } catch {
            print("[Auth] Logout all devices API call failed: \(error)")
        }

        tokenStore.clearAll()
        currentUser = nil
        isAuthenticated = false
        authState = .unauthenticated
    }

    // MARK: - Token Access

    /// Get a valid access token, refreshing if needed
    func getAccessToken() async throws -> String {
        if tokenStore.isAccessTokenValid, let token = tokenStore.accessToken {
            return token
        }

        guard let refreshToken = tokenStore.refreshToken else {
            throw AuthError.notAuthenticated
        }

        let response = try await api.refreshTokens(refreshToken)
        tokenStore.accessToken = response.accessToken
        tokenStore.accessTokenExpiry = Date().addingTimeInterval(TimeInterval(response.expiresIn))
        tokenStore.refreshToken = response.refreshToken

        guard let token = tokenStore.accessToken else {
            throw AuthError.invalidToken
        }

        return token
    }
}

File: win_app/src/QueenMama.Core/Services/Auth/AuthenticationManager.cs
using CommunityToolkit.Mvvm.ComponentModel;
using Microsoft.Extensions.Logging;
using QueenMama.Core.Interfaces;
using QueenMama.Core.Models;

namespace QueenMama.Core.Services.Auth;

/// <summary>
/// Main authentication coordinator
/// Manages login flows, token refresh, and authentication state
/// </summary>
public partial class AuthenticationManager : ObservableObject, IAuthenticationManager
{
    private readonly IAuthApiClient _api;
    private readonly AuthTokenStore _tokenStore;
    private readonly ILogger<AuthenticationManager> _logger;

    private CancellationTokenSource? _pollingCts;
    private const int PollingIntervalSeconds = 5;

    public event Action<AuthState>? OnAuthStateChanged;
    public event Action<Exception>? OnError;

    [ObservableProperty]
    private AuthState _currentState = AuthState.Unknown;

    [ObservableProperty]
    private User? _currentUser;

    [ObservableProperty]
    private bool _isAuthenticated;

    public AuthenticationManager(
        IAuthApiClient api,
        IAuthTokenStore tokenStore,
        ILogger<AuthenticationManager> logger)
    {
        _api = api;
        _tokenStore = (AuthTokenStore)tokenStore;
        _logger = logger;
    }

    partial void OnCurrentStateChanged(AuthState value)
    {
        OnAuthStateChanged?.Invoke(value);
    }

    /// <summary>
    /// Check for existing authentication on app launch
    /// </summary>
    public async Task<bool> CheckExistingAuthAsync()
    {
        CurrentState = AuthState.Unknown;

        // Check for stored credentials
        if (!await _tokenStore.HasStoredCredentialsAsync())
        {
            CurrentState = AuthState.SignedOut;
            return false;
        }

        var storedUser = await _tokenStore.GetStoredUserAsync();
        if (storedUser == null)
        {
            CurrentState = AuthState.SignedOut;
            return false;
        }

        // Try to validate/refresh tokens
        try
        {
            if (!_tokenStore.IsAccessTokenValid)
            {
                var refreshToken = await _tokenStore.GetRefreshTokenAsync();
                if (refreshToken == null)
                {
                    throw new AuthException(AuthErrorType.TokenExpired);
                }

                var result = await _api.RefreshTokenAsync(refreshToken);
                if (!result.Success || result.AccessToken == null)
                {
                    throw new AuthException(AuthErrorType.TokenExpired);
                }

                _tokenStore.SetAccessToken(result.AccessToken, result.ExpiresAt ?? DateTime.UtcNow.AddHours(1));
                if (result.RefreshToken != null)
                {
                    await _tokenStore.SetRefreshTokenAsync(result.RefreshToken);
                }
            }

            // Successfully authenticated
            CurrentUser = storedUser;
            IsAuthenticated = true;
            CurrentState = AuthState.Authenticated;

            _logger.LogInformation("Existing auth validated for {Email}", storedUser.Email);
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Existing auth validation failed");
            await _tokenStore.ClearTokensAsync();
            CurrentState = AuthState.SignedOut;
            return false;
        }
    }

    /// <summary>
    /// Start the device code flow for OAuth users
    /// </summary>
    public async Task<DeviceCodeResponse> StartDeviceCodeFlowAsync()
    {
        CurrentState = AuthState.Authenticating;

        try
        {
            var response = await _api.RequestDeviceCodeAsync();
            _logger.LogInformation("Device code flow started, user code: {UserCode}", response.UserCode);

            // Start polling in background
            StartPolling(response.DeviceCode, response.ExpiresAt);

            return response;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to start device code flow");
            CurrentState = AuthState.Error;
            OnError?.Invoke(ex);
            throw;
        }
    }

    /// <summary>
    /// Poll for authorization completion
    /// </summary>
    public async Task<AuthResult> PollForAuthorizationAsync(string deviceCode, CancellationToken cancellationToken = default)
    {
        try
        {
            var result = await _api.PollDeviceCodeAsync(deviceCode);

            if (result.Success && result.User != null && result.AccessToken != null)
            {
                // Store tokens
                var tokens = new AuthTokens
                {
                    AccessToken = result.AccessToken,
                    RefreshToken = result.RefreshToken ?? "",
                    ExpiresIn = (int)(result.ExpiresAt ?? DateTime.UtcNow.AddHours(1)).Subtract(DateTime.UtcNow).TotalSeconds
                };

                await _tokenStore.StoreTokensAndUserAsync(tokens, result.User);

                // Update state
                CurrentUser = result.User;
                IsAuthenticated = true;
                CurrentState = AuthState.Authenticated;

                _logger.LogInformation("Authentication successful for {Email}", result.User.Email);
            }

            return result;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Polling failed");
            throw;
        }
    }

    private void StartPolling(string deviceCode, DateTime expiresAt)
    {
        _pollingCts?.Cancel();
        _pollingCts = new CancellationTokenSource();

        _ = Task.Run(async () =>
        {
            while (!_pollingCts.Token.IsCancellationRequested && DateTime.UtcNow < expiresAt)
            {
                try
                {
                    await Task.Delay(TimeSpan.FromSeconds(PollingIntervalSeconds), _pollingCts.Token);

                    if (_pollingCts.Token.IsCancellationRequested)
                        break;

                    var result = await PollForAuthorizationAsync(deviceCode, _pollingCts.Token);

                    if (result.Success)
                    {
                        return;
                    }

                    // Continue polling if authorization is pending
                    if (result.Error == "authorization_pending")
                    {
                        continue;
                    }

                    // Some other error
                    if (result.Error != null && result.Error != "authorization_pending")
                    {
                        CurrentState = AuthState.Error;
                        OnError?.Invoke(new AuthException(AuthErrorType.ServerError, result.Message));
                        return;
                    }
                }
                catch (TaskCanceledException)
                {
                    break;
                }
                catch (Exception ex)
                {
                    _logger.LogWarning(ex, "Polling error, continuing...");
                }
            }

            // Expired without authorization
            if (!_pollingCts.Token.IsCancellationRequested && CurrentState != AuthState.Authenticated)
            {
                CurrentState = AuthState.Error;
                OnError?.Invoke(new AuthException(AuthErrorType.ServerError, "Device code expired. Please try again."));
            }
        });
    }

    /// <summary>
    /// Cancel the device code flow
    /// </summary>
    public void CancelDeviceCodeFlow()
    {
        _pollingCts?.Cancel();
        _pollingCts = null;
        CurrentState = AuthState.SignedOut;
    }

    /// <summary>
    /// Refresh the access token
    /// </summary>
    public async Task<bool> RefreshTokenAsync()
    {
        var refreshToken = await _tokenStore.GetRefreshTokenAsync();
        if (refreshToken == null)
        {
            _logger.LogWarning("No refresh token available");
            return false;
        }

        try
        {
            var result = await _api.RefreshTokenAsync(refreshToken);
            if (result.Success && result.AccessToken != null)
            {
                _tokenStore.SetAccessToken(result.AccessToken, result.ExpiresAt ?? DateTime.UtcNow.AddHours(1));
                if (result.RefreshToken != null)
                {
                    await _tokenStore.SetRefreshTokenAsync(result.RefreshToken);
                }
                return true;
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Token refresh failed");
        }

        return false;
    }

    /// <summary>
    /// Sign out from current device
    /// </summary>
    public async Task SignOutAsync()
    {
        _pollingCts?.Cancel();

        var refreshToken = await _tokenStore.GetRefreshTokenAsync();
        if (refreshToken != null)
        {
            try
            {
                await _api.RevokeTokenAsync(refreshToken);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to revoke token on server");
            }
        }

        await _tokenStore.ClearTokensAsync();
        CurrentUser = null;
        IsAuthenticated = false;
        CurrentState = AuthState.SignedOut;

        _logger.LogInformation("User signed out");
    }

    /// <summary>
    /// Validate the current session
    /// </summary>
    public async Task<bool> ValidateSessionAsync()
    {
        if (!IsAuthenticated)
            return false;

        if (_tokenStore.IsAccessTokenValid)
            return true;

        return await RefreshTokenAsync();
    }

    /// <summary>
    /// Get a valid access token
    /// </summary>
    public async Task<string?> GetAccessTokenAsync()
    {
        if (_tokenStore.IsAccessTokenValid && _tokenStore.AccessToken != null)
        {
            return _tokenStore.AccessToken;
        }

        if (await RefreshTokenAsync())
        {
            return _tokenStore.AccessToken;
        }

        return null;
    }
}
